### PR TITLE
build: rebuild if git HEAD changed

### LINKED ISSUE #431: version SHA for runtime doesn't update unless make clean
after making a local commit during development, I expect the SHA associated with cc-runtime --version to be updated after doing a make.  This seems to only occur after doing a make clean first.  Note that the SHA exhibited during the build process doesn't match the resulting version given with cc-runtime --version:

```
eernst@eernst-nuc-2:~/go/src/github.com/clearcontainers/runtime$ ./cc-runtime --version
cc-runtime  : 3.0.0-beta.1
   commit   : 6ea23563d970c57f12c64f7ca40bbb817dfdca5d
   OCI specs: 1.0.0-rc5
eernst@eernst-nuc-2:~/go/src/github.com/clearcontainers/runtime$ git log -n1 --oneline
6ea2356 Revert "runtime: Switch to the new agent"
# Up until this point, all is well. Next, make a trivial change to update SHA:
#
eernst@eernst-nuc-2:~/go/src/github.com/clearcontainers/runtime$ git commit --amend -s
[master 3a0a091] Revert "runtime: Switch to the new agent"
 Date: Wed Aug 16 15:51:00 2017 -0700
 2 files changed, 1 insertion(+), 2 deletions(-)
eernst@eernst-nuc-2:~/go/src/github.com/clearcontainers/runtime$ git log -n1 --oneline
3a0a091 Revert "runtime: Switch to the new agent"
eernst@eernst-nuc-2:~/go/src/github.com/clearcontainers/runtime$ make
cc-runtime - version 3.0.0-beta.1 (commit 3a0a091f3777411c1262cf69e755ad800b895188)

• Summary:

        Clear Containers system build     : no

        binary install path (DESTTARGET)  : /usr/local/bin/cc-runtime
        config install path (DESTCONFIG)  : /usr/local/etc/clear-containers/configuration.toml
        hypervisor path (QEMUPATH)        : /usr/local/bin/qemu-lite-system-x86_64
        assets path (PKGDATADIR)          : /usr/local/share/clear-containers
        proxy+shim path (PKGLIBEXECDIR)   : /usr/local/libexec/clear-containers
        pause bundle path (PAUSEROOTPATH) : /usr/local/var/lib/clear-containers/runtime/bundles/pause_bundle
#Note build output above matches what is expected for source SHA, while version
# provided below is the prior HEAD
#
eernst@eernst-nuc-2:~/go/src/github.com/clearcontainers/runtime$ ./cc-runtime --version
cc-runtime  : 3.0.0-beta.1
   commit   : 6ea23563d970c57f12c64f7ca40bbb817dfdca5d
   OCI specs: 1.0.0-rc5
eernst@eernst-nuc-2:~/go/src/github.com/clearcontainers/runtime$ make clean
     CLEAN   clean
     CLEAN   clean
eernst@eernst-nuc-2:~/go/src/github.com/clearcontainers/runtime$ make
     GENERATE config-generated.go
cc-runtime - version 3.0.0-beta.1 (commit 3a0a091f3777411c1262cf69e755ad800b895188)

• Summary:

        Clear Containers system build     : no

        binary install path (DESTTARGET)  : /usr/local/bin/cc-runtime
        config install path (DESTCONFIG)  : /usr/local/etc/clear-containers/configuration.toml
        hypervisor path (QEMUPATH)        : /usr/local/bin/qemu-lite-system-x86_64
        assets path (PKGDATADIR)          : /usr/local/share/clear-containers
        proxy+shim path (PKGLIBEXECDIR)   : /usr/local/libexec/clear-containers
        pause bundle path (PAUSEROOTPATH) : /usr/local/var/lib/clear-containers/runtime/bundles/pause_bundle

     BUILD   cc-runtime
     CONFIG  config/configuration.toml
cc -Os -Wall -Wextra -static   -c -o pause/pause.o pause/pause.c
     BUILD   pause
eernst@eernst-nuc-2:~/go/src/github.com/clearcontainers/runtime$ ./cc-runtime --version
cc-runtime  : 3.0.0-beta.1
   commit   : 3a0a091f3777411c1262cf69e755ad800b895188
   OCI specs: 1.0.0-rc5
```

### PR DESCRIPTION
**Pull request by dvoytik**
Install git hooks which handle cases when git HEAD is changed.
The git hooks just remove config-generated.go file to force
binary rebuild.

Fixes #431.

Signed-off-by: Dmitry Voytik <dmitry.voytik@huawei.com>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -227,7 +227,8 @@ pause: pause/pause.go
 	pause \
 	show-header \
 	show-summary \
-	show-variables
+	show-variables \
+	install-git-hooks
```
Comment (grahamwhaley) [inline Makefile]: This list looks to be alphabetical - can you move your addition up to maintain that?

File: Makefile [hunk 2]
```diff
@@ -328,3 +329,43 @@ show-summary: show-header
...
+
+# This git-hook is executed after every checkout git operation
+.git/hooks/post-checkout: Makefile
+	@ mkdir -p .git/hooks/
+	$(QUIET_INST)echo "$$GIT_HOOK_POST_CHECKOUT" >$@
+	@ chmod +x $@
+
+# This git-hook is executed after every commit, merge, amend or rebase git
+# operation
+.git/hooks/post-commit .git/hooks/post-merge .git/hooks/post-rewrite: Makefile
```
Comment (grahamwhaley) [inline Makefile]: multi target rule, with files beginning with `'` - that made me squint a bit :-)

**Comment by grahamwhaley** [comment]
lgtm

**Comment by grahamwhaley** [comment]
Adding @jodh-intel as reviewer for another pair of eyes :-)

**Comment by grahamwhaley** [comment]
I guess I'll add a thought for @jodh-intel to mull on, as I'm not sure and I didn't test.... are we currently doing 'source releases' of this repo - in which case, does this still hang together when there is no `.git` dir (as we mentioned before in the chatter)

File: Makefile (hunk 1 above)
Comment (dvoytik) [inline Makefile]: Thanks! Will do.

File: Makefile (hunk 2 above)
Comment (dvoytik) [inline Makefile]: Would be it better this way:
```
.git/hooks/post-commit \
.git/hooks/post-merge \
.git/hooks/post-rewrite: Makefile
        # ...
```

Comment (grahamwhaley) [inline Makefile]: no - it's fine as it is. Just my eyes are not used to either multi lhs targets or targets starting with a `.` - so the combination was even more fun :-)

**Comment by dvoytik** [comment]
@grahamwhaley, thank you for the review! I've just tested after removing .git and it works without problems.

**Pull request merged**